File: packages/core/src/explain.ts
// ─── Explanation Helpers ────────────────────────────────────────────────────
// Shared utilities for source-backed calculator explanations.

import type { RateSourceMetadata } from './types';
import { explain as explainRate } from './rates';

export interface SourceCollection {
  sources: RateSourceMetadata[];
  warnings: string[];
}

export function uniqueRateKeys(rateKeys: string[]): string[] {
  return [...new Set(rateKeys)];
}

export function collectRateSources(rateKeys: string[]): SourceCollection {
  const sources: RateSourceMetadata[] = [];
  const warnings: string[] = [];

  for (const key of uniqueRateKeys(rateKeys)) {
    try {
      const source = explainRate(key);
      sources.push(source);
      warnings.push(...source.warnings);
      if (source.verificationStatus !== 'verified') {
        warnings.push(`${key}: source status is ${source.verificationStatus} with ${source.confidence} confidence. ${source.notes}`);
      }
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      warnings.push(`${key}: source metadata unavailable. ${message}`);
    }
  }

  return { sources, warnings };
}

File: packages/python/src/ngtaxkit/explain.py
"""Shared helpers for source-backed calculation explanations."""

from __future__ import annotations

from . import rates
from .types import RateSourceMetadata


def unique_rate_keys(rate_keys: list[str]) -> list[str]:
    """Return rate keys in first-seen order without duplicates."""
    return list(dict.fromkeys(rate_keys))


def collect_rate_sources(rate_keys: list[str]) -> tuple[list[RateSourceMetadata], list[str]]:
    """Collect source records for rate keys and warnings for weak or missing sources."""
    sources: list[RateSourceMetadata] = []
    warnings: list[str] = []

    for key in unique_rate_keys(rate_keys):
        try:
            source = rates.explain(key)
        except Exception as exc:
            warnings.append(f"{key}: source metadata unavailable. {exc}")
            continue

        sources.append(source)
        if source["verification_status"] != "verified":
            warnings.append(
                f'{key}: source status is {source["verification_status"]} '
                f'with {source["confidence"]} confidence. {source["notes"]}'
            )

    return sources, warnings
